src/UserSpaceInstrumentation/AccessTraceesMemory.cpp

```diff
@@ -25,44 +26,61 @@ using orbit_base::ReadFileToString;
 [[nodiscard]] ErrorMessageOr<std::vector<uint8_t>> ReadTraceesMemory(pid_t pid,
                                                                      uint64_t address_start,
                                                                      uint64_t length) {
-  // Round up length to next multiple of eight.
-  length = ((length + 7) / 8) * 8;
-  std::vector<uint8_t> bytes(length);
-  for (size_t i = 0; i < length / 8; i++) {
-    const uint64_t data = ptrace(PTRACE_PEEKDATA, pid, address_start + i * 8, NULL);
-    if (errno) {
-      return ErrorMessage(
-          absl::StrFormat("Failed to PTRACE_PEEKDATA for pid %d with errno %d: \"%s\"", pid, errno,
-                          SafeStrerror(errno)));
-    }
-    std::memcpy(bytes.data() + (i * sizeof(uint64_t)), &data, 8);
+  if (length == 0) {
+    return ErrorMessage("Tried to read 0 bytes.");
   }
-  return bytes;
+
+  OUTCOME_TRY(fd, orbit_base::OpenFileForReading(absl::StrFormat("/proc/%d/mem", pid)));
+
+  std::vector<uint8_t> data(length);
+  ssize_t result = pread(fd.get(), data.data(), length, address_start);
+
+  if (result == -1) {
+    return ErrorMessage(
+        absl::StrFormat("Failed to read from memory file of process %d with errno %d: \"%s\"", pid,
+                        errno, SafeStrerror(errno)));
+  }
+
+  if (result < static_cast<ssize_t>(length)) {
+    return ErrorMessage(absl::StrFormat(
+        "Failed to read %u bytes from memory file of process %d. Only got %d bytes.", length, pid,
+        result));
+  }
+
+  return data;
 }
 
 [[nodiscard]] ErrorMessageOr<void> WriteTraceesMemory(pid_t pid, uint64_t address_start,
                                                       const std::vector<uint8_t>& bytes) {
-  size_t pos = 0;
-  do {
-    // Pack 8 bytes for writing into `data`.
-    uint64_t data = 0;
-    std::memcpy(&data, bytes.data() + pos, sizeof(uint64_t));
-    if (ptrace(PTRACE_POKEDATA, pid, address_start + pos, data) == -1) {
-      return ErrorMessage(absl::StrFormat("Failed to PTRACE_POKEDATA for pid: %d.", pid));
-    }
-    pos += sizeof(uint64_t);
-  } while (pos < bytes.size());
+  if (bytes.empty()) {
+    return ErrorMessage("Tried to write 0 bytes.");
+  }
+  OUTCOME_TRY(fd, orbit_base::OpenFileForWriting(absl::StrFormat("/proc/%d/mem", pid)));
+  ssize_t result = pwrite(fd.get(), bytes.data(), bytes.size(), address_start);
```

Comment:
same here - WriteFullyAtOffset, same thing about having written less than length is not a error.